### ISSUE TITLE
JDK8 engine not supported?

### ISSUE THREAD

**Issue by ScalaWilliam** [nidi3/graphviz-java#21]
I see this commit here:
https://github.com/nidi3/graphviz-java/commit/c5dd1010fcfc75cdae06a147a67011f3db273798#diff-401061ebf2e550ad9f320a6b6f32e41eR35

I removed the throw statement and then I tried rendering an SVG on Nashown 1.8.0_111 - it was fine.

What is the reason for this? Is there a reproducible test case I can help with?

**Comment by nidi3** [comment]
Hi there
Thanks for the message. It's interesting, I tried to run `EngineTest.jdk` on MacOS with versions 1.8.0_77, 91, 101, 111, 121, 131 and I always get this exception. 
I'll remove the check, catch this exception and log an appropriate message.
```
Exception in thread "Thread-1" java.lang.AssertionError: Failed generating bytecode for <eval>:1156
	at jdk.nashorn.internal.codegen.CompilationPhase$BytecodeGenerationPhase.transform(CompilationPhase.java:431)
	at jdk.nashorn.internal.codegen.CompilationPhase.apply(CompilationPhase.java:624)
	at jdk.nashorn.internal.codegen.Compiler.compile(Compiler.java:655)
	at jdk.nashorn.internal.runtime.RecompilableScriptFunctionData.compileTypeSpecialization(RecompilableScriptFunctionData.java:725)
	at jdk.nashorn.internal.runtime.RecompilableScriptFunctionData.getBest(RecompilableScriptFunctionData.java:905)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.getBest(ScriptFunctionData.java:375)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.getBestInvoker(ScriptFunctionData.java:237)
	at jdk.nashorn.internal.runtime.ScriptFunction.findCallMethod(ScriptFunction.java:871)
	at jdk.nashorn.internal.runtime.ScriptObject.lookup(ScriptObject.java:1825)
	at jdk.nashorn.internal.runtime.linker.NashornLinker.getGuardedInvocation(NashornLinker.java:104)
	at jdk.nashorn.internal.runtime.linker.NashornLinker.getGuardedInvocation(NashornLinker.java:98)
	at jdk.internal.dynalink.support.CompositeTypeBasedGuardingDynamicLinker.getGuardedInvocation(CompositeTypeBasedGuardingDynamicLinker.java:176)
	at jdk.internal.dynalink.support.CompositeGuardingDynamicLinker.getGuardedInvocation(CompositeGuardingDynamicLinker.java:124)
	at jdk.internal.dynalink.support.LinkerServicesImpl.getGuardedInvocation(LinkerServicesImpl.java:154)
	at jdk.internal.dynalink.DynamicLinker.relink(DynamicLinker.java:253)
	at jdk.nashorn.internal.scripts.Script$Recompilation$165$3514339III$\^eval\_.L:22#Module#Y#P3(<eval>:1156)
	at jdk.nashorn.internal.scripts.Script$Recompilation$164$733939III$\^eval\_.L:22#Module#Y#xF(<eval>:1145)
	at jdk.nashorn.internal.scripts.Script$Recompilation$163$992196I$\^eval\_.L:22#Module#Y#pS(<eval>:1146)
	at jdk.nashorn.internal.scripts.Script$Recompilation$161$991369II$\^eval\_.L:22#Module#Y#mS(<eval>:1146)
	at jdk.nashorn.internal.scripts.Script$Recompilation$96$753593$\^eval\_.L:22#Module#Y#_F(<eval>:1145)
	at jdk.nashorn.internal.scripts.Script$Recompilation$95$\^eval\_.L:22#Module#Y#dc(<eval>:1147)
	at jdk.nashorn.internal.scripts.Script$Recompilation$87$5816AAAAA$\^eval\_.L:22#Module#L:37#ra(<eval>:37)
	at jdk.nashorn.internal.scripts.Script$Recompilation$5$3608716AAA$\^eval\_.L:22#render(<eval>:1204)
	at jdk.nashorn.internal.scripts.Script$Recompilation$4$3608102A$\^eval\_.L:22#Viz(<eval>:1186)
	at jdk.nashorn.internal.scripts.Script$3$\^eval\_.:program(<eval>:1)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:449)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:406)
	at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:402)
	at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at guru.nidi.graphviz.engine.GraphvizJdkEngine.doInit(GraphvizJdkEngine.java:65)
	at guru.nidi.graphviz.engine.AbstractGraphvizEngine.init(AbstractGraphvizEngine.java:56)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.AssertionError
	at jdk.nashorn.internal.codegen.Label$Stack.mergeVariableTypes(Label.java:187)
	at jdk.nashorn.internal.codegen.Label$Stack.joinFrom(Label.java:168)
	at jdk.nashorn.internal.codegen.Label.joinFrom(Label.java:548)
	at jdk.nashorn.internal.codegen.MethodEmitter.joinTo(MethodEmitter.java:1884)
	at jdk.nashorn.internal.codegen.MethodEmitter.breakLabel(MethodEmitter.java:1906)
	at jdk.nashorn.internal.codegen.MethodEmitter.label(MethodEmitter.java:1892)
	at jdk.nashorn.internal.codegen.CodeGenerator.enterIfNode(CodeGenerator.java:2190)
	at jdk.nashorn.internal.ir.IfNode.accept(IfNode.java:86)
	at jdk.nashorn.internal.ir.Node.accept(Node.java:271)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:185)
	at jdk.nashorn.internal.ir.LexicalContextNode$Acceptor.accept(LexicalContextNode.java:57)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:439)
	at jdk.nashorn.internal.codegen.CodeGenerator.enterDoWhile(CodeGenerator.java:3469)
	at jdk.nashorn.internal.codegen.CodeGenerator.enterWhileNode(CodeGenerator.java:3377)
	at jdk.nashorn.internal.ir.WhileNode.accept(WhileNode.java:82)
	at jdk.nashorn.internal.ir.LexicalContextNode$Acceptor.accept(LexicalContextNode.java:57)
	at jdk.nashorn.internal.ir.LexicalContextStatement.accept(LexicalContextStatement.java:55)
	at jdk.nashorn.internal.ir.LoopNode.accept(LoopNode.java:36)
	at jdk.nashorn.internal.ir.Node.accept(Node.java:271)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:185)
	at jdk.nashorn.internal.ir.LexicalContextNode$Acceptor.accept(LexicalContextNode.java:57)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:439)
	at jdk.nashorn.internal.ir.LabelNode.accept(LabelNode.java:80)
	at jdk.nashorn.internal.ir.LexicalContextNode$Acceptor.accept(LexicalContextNode.java:57)
	at jdk.nashorn.internal.ir.LexicalContextStatement.accept(LexicalContextStatement.java:55)
	at jdk.nashorn.internal.ir.LabelNode.accept(LabelNode.java:36)
	at jdk.nashorn.internal.ir.Node.accept(Node.java:271)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:185)
	at jdk.nashorn.internal.ir.LexicalContextNode$Acceptor.accept(LexicalContextNode.java:57)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:439)
	at jdk.nashorn.internal.codegen.CodeGenerator.enterIfNode(CodeGenerator.java:2186)
	at jdk.nashorn.internal.ir.IfNode.accept(IfNode.java:86)
	at jdk.nashorn.internal.ir.Node.accept(Node.java:271)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:185)
	at jdk.nashorn.internal.ir.LexicalContextNode$Acceptor.accept(LexicalContextNode.java:57)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:439)
	at jdk.nashorn.internal.codegen.CodeGenerator.enterIfNode(CodeGenerator.java:2186)
	at jdk.nashorn.internal.ir.IfNode.accept(IfNode.java:86)
	at jdk.nashorn.internal.ir.Node.accept(Node.java:271)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:185)
	at jdk.nashorn.internal.ir.LexicalContextNode$Acceptor.accept(LexicalContextNode.java:57)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:439)
	at jdk.nashorn.internal.codegen.CodeGenerator.enterDoWhile(CodeGenerator.java:3469)
	at jdk.nashorn.internal.codegen.CodeGenerator.enterWhileNode(CodeGenerator.java:3377)
	at jdk.nashorn.internal.ir.WhileNode.accept(WhileNode.java:82)
	at jdk.nashorn.internal.ir.LexicalContextNode$Acceptor.accept(LexicalContextNode.java:57)
	at jdk.nashorn.internal.ir.LexicalContextStatement.accept(LexicalContextStatement.java:55)
	at jdk.nashorn.internal.ir.LoopNode.accept(LoopNode.java:36)
	at jdk.nashorn.internal.ir.Node.accept(Node.java:271)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:185)
	at jdk.nashorn.internal.ir.LexicalContextNode$Acceptor.accept(LexicalContextNode.java:57)
	at jdk.nashorn.internal.ir.Block.accept(Block.java:439)
	at jdk.nashorn.internal.ir.FunctionNode.accept(FunctionNode.java:336)
	at jdk.nashorn.internal.ir.LexicalContextNode$Acceptor.accept(LexicalContextNode.java:57)
	at jdk.nashorn.internal.ir.LexicalContextExpression.accept(LexicalContextExpression.java:47)
	at jdk.nashorn.internal.ir.FunctionNode.accept(FunctionNode.java:55)
	at jdk.nashorn.internal.codegen.CompilationPhase.transformFunction(CompilationPhase.java:628)
	at jdk.nashorn.internal.codegen.CompilationPhase.access$100(CompilationPhase.java:55)
	at jdk.nashorn.internal.codegen.CompilationPhase$BytecodeGenerationPhase.transform(CompilationPhase.java:418)
	... 35 more
```

**Comment by ScalaWilliam** [comment]
That test passed for me, interestingly enough.

**Comment by ScalaWilliam** [comment]
Thanks very much! I will see if I can reproduce perhaps. This library is very cool :-)

**Comment by ScalaWilliam** [comment]
```
$ mvn -Dtest=EngineTest test
[INFO] Running guru.nidi.graphviz.engine.EngineTest
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 26.37 s - in guru.nidi.graphviz.engine.EngineTest
...
$ java -version
java version "1.8.0_111"
Java(TM) SE Runtime Environment (build 1.8.0_111-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.111-b14, mixed mode)
$ system_profiler SPSoftwareDataType|grep Version
      System Version: macOS 10.12.5 (16F73)
      Kernel Version: Darwin 16.6.0
```

**Comment by nidi3** [comment]
Ah! Found it! 
It's the -ea JVM option. It enables asserts. It's set by default by IntelliJ, if I remove it, the test passes.
Learned something again.

**Comment by nidi3** [comment]
Now asserts are diabled for just the problematic class. The issue should be solved with this.

**Comment by nidi3** [comment]
Thanks for the persistence.

**Comment by ScalaWilliam** [comment]
Ah! I found another problem, while trying to use the latest viz js. 

See the description here: https://github.com/ScalaWilliam/fail-nashorn-test

TLDR trying inside graphviz-java with latest viz gives a VerifyError but trying in a separate project works just fine. Not sure what to do about this.

**Comment by nidi3** [comment]
Tried viz-1.8.0 and works both with nashorn and v8...

**Comment by ScalaWilliam** [comment]
Excellent! Maybe a cache invalidation needed somewhere or so.

Thanks for fast response :)
